Sources/WireGuardKitGo/tunnel-handle.go

```diff
@@ -0,0 +1,239 @@
+}
+
+func NewTunnelHandle(exit *device.Device, entry *device.Device, logger *device.Logger, virtualNet *netstack.Net) tunnelHandle {
+	return tunnelHandle{
+		exit:          exit,
+		entry:         entry,
+		logger:        logger,
+		VirtualNet:    virtualNet,
+		socketHandles: make(map[int32]*socketHandle),
+		lock:          &sync.Mutex{},
+	}
+}
+
+// Returns nil if tunnel is closed
+func (tun *tunnelHandle) GetConfig() *string {
+	settings, err := tun.exit.IpcGet()
+	if err != nil {
+		return nil
+	}
+	return &settings
+}
+
+func (tun *tunnelHandle) SetConfig(settings string) int64 {
+	err := tun.exit.IpcSet(settings)
+	if err != nil {
+		tun.logger.Errorf("Unable to set IPC settings: %v", err)
+		if ipcErr, ok := err.(*device.IPCError); ok {
+			return ipcErr.ErrorCode()
+		}
+		return errBadWgConfig
+	}
+	return 0
+}
+
+func (tun *tunnelHandle) BumpSockets() {
+	device := tun.exit
+	if tun.entry != nil {
+		device = tun.entry
+	}
+
+	go func() {
+		for i := 0; i < 10; i++ {
+			err := device.BindUpdate()
+			if err == nil {
+				device.SendKeepalivesToPeersWithCurrentKeypair()
+				return
+			}
+			tun.logger.Errorf("Unable to update bind, try %d: %v", i+1, err)
+			time.Sleep(time.Second / 2)
+		}
+		tun.logger.Errorf("Gave up trying to update bind; tunnel is likely dysfunctional")
+	}()
+}
+
+func (tun *tunnelHandle) DisableSomeRoamingForBrokenMobileSemantics() {
+	tun.exit.DisableSomeRoamingForBrokenMobileSemantics()
+	if tun.entry != nil {
+		tun.entry.DisableSomeRoamingForBrokenMobileSemantics()
+	}
+}
+
+// Creates a socket asynchronously and returns an handle to it immediately.
+// Calls to get the socket will block until the passed in closure returns. The
+// closure takes a context and the virtual networking stack. Any connection
+// returned from the closure should be bound to virtual network.
+func (tun *tunnelHandle) AddSocket(ctx context.Context, createSocket func(ctx context.Context, virtualNet *netstack.Net) (net.Conn, error)) int32 {
+	tun.lock.Lock()
+	defer tun.lock.Unlock()
+
+	socketHandle := newSocketHandle(tun.VirtualNet, ctx, createSocket)
+	handle := insertHandle(tun.socketHandles, socketHandle)
+	// Whilst technically we could try getting an unused key into the map
+	// before creating a handle, it is far too unlikely that we will run out of
+	// int32 handles that the incurred mess of that is not worth it.
+	if handle < 0 {
+		socketHandle.close()
+	}
+	return handle
+}
+
+// Returns a socket bound to the virtual network. Blocks until socket is connected.
+func (tun *tunnelHandle) GetSocket(id int32) (net.Conn, error, bool) {
+	tun.lock.Lock()
+	socket, ok := tun.socketHandles[id]
+	tun.lock.Unlock()
+
+	if !ok {
+		return nil, nil, false
+	}
+
+	conn, err := socket.Get()
+
+	return conn, err, true
+}
+
+func (tun *tunnelHandle) RemoveAndCloseSocket(id int32) bool {
+	tun.lock.Lock()
+	defer tun.lock.Unlock()
+	socket, ok := tun.socketHandles[id]
+	if ok {
+		socket.close()
+	}
+
+	delete(tun.socketHandles, id)
```

Comment:
Should the entry be deleted even if it was not present?

Reply:
Nope, this will probably panic.  Will fix.
<!-- Reviewable comment -ODfvM7A52jvICFAdR95:bry4rce -->
<!-- Sent from Reviewable.io -->